puma/utils/discriminant.py

```diff
@@ -0,0 +1,145 @@
+"""Discriminant calculation for flavour tagging."""
+import numpy as np
+
+from puma.utils.histogram import save_divide
+
+
+def calc_disc(
+    scores: np.ndarray, flvs: list = None, flv_map: dict = None
+) -> np.ndarray:
+    """Calculate arbitrary flavour tagging score.
+
+    Parameters
+    ----------
+    scores : np.ndarray
+        tagger scores in the shape (n_jets, n_flavours)
+    flvs : list, optional
+        , by default None
+    flv_map : dict, optional
+        flavour map containing signal and background mapping from `flv` and their
+        fractions, by default None
+
+    Returns
+    -------
+    np.ndarray
+        discriminant values
+
+    Raises
+    ------
+    ValueError
+        if scores and shapes have different shapes
+
+    Examples
+    --------
+    The `flv_map` can e.g. look like this together with the `flvs`
+    >>> flv_map ={
+    ...     "sig": {"b": 1.0},
+    ...     "bkg": {"l": 1 - 0.5, "c": 0.5},
+    ... }
+    >>> flvs = ["l", "c", "b"]
+    >>> scores = np.column_stack((np.ones(10), np.ones(10), np.ones(10)))
+    >>> calc_disc(scores, flvs=flvs, flv_map=flv_map)
+    array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])
+    """
+    flvs = ["l", "c", "b"] if flvs is None else flvs
+    if len(flvs) != scores.shape[1]:
+        raise ValueError("`flvs` and `scores` have incompatible shapes.")
+    flv_map = (
+        {
+            "sig": {"b": 1.0},
+            "bkg": {"l": 1 - 0.5, "c": 0.5},
+        }
+        if flv_map is None
+        else flv_map
+    )
+    numerator = None
+    for sig_name, frac in flv_map["sig"].items():
+        sig_arr = scores[:, flvs.index(sig_name)]
+        numerator = frac * sig_arr if numerator is None else numerator + frac * sig_arr
+    denominator = None
+    for bkg_name, frac in flv_map["bkg"].items():
+        bkg_arr = scores[:, flvs.index(bkg_name)]
+        denominator = (
+            frac * bkg_arr if denominator is None else denominator + frac * bkg_arr
+        )
+    discs = np.log(
+        save_divide(
+            numerator,
+            denominator,
+            default=np.infty,
+        )
+    )
+    return discs
+
+
+def calc_disc_b(
+    arr_pu: np.ndarray, arr_pc: np.ndarray, arr_pb: np.ndarray, fc_par: float
```

Comment:
```suggestion
    arr_pu: np.ndarray,
    arr_pc: np.ndarray,
    arr_pb: np.ndarray,
    fc_par: float,
```

Reply:
And also, should we hardcode pu, pb, pc?

Reply:
that's a special case function, there is a new function `calc_disc` which can handle signal and bkg more dynamically